services/scheduler/src/main/java/org/mule/service/scheduler/internal/DefaultScheduler.java

```diff
@@ -0,0 +1,282 @@
+            task.cancel(true);
+          }
+          if (!task.isStarted()) {
+            tasks.add(task);
+          }
+        }
+        currentTasks.clear();
+
+        return tasks;
+      }
+    } finally {
+      tryTerminate();
+    }
+  }
+
+  @Override
+  public boolean isShutdown() {
+    return shutdown || executor.isShutdown();
+  }
+
+  @Override
+  public boolean isTerminated() {
+    return shutdown && currentTasks.isEmpty();
+  }
+
+  @Override
+  public boolean awaitTermination(long timeout, TimeUnit unit) throws InterruptedException {
+    final long stopNanos = nanoTime() + unit.toNanos(timeout);
+    while (nanoTime() <= stopNanos) {
+      if (isTerminated()) {
+        return true;
+      }
+      // Do this in a while just in case that the termination occurred right before this next line is executed.
+      if (terminationLatch.await(50, MILLISECONDS)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  @Override
+  public void stop(long gracefulShutdownTimeoutSecs, TimeUnit unit) {
+    // Disable new tasks from being submitted
+    shutdown();
+    try {
+      // Wait a while for existing tasks to terminate
+      if (!awaitTermination(gracefulShutdownTimeoutSecs, unit)) {
+        // Cancel currently executing tasks and return list of pending
+        // tasks
+        List<Runnable> cancelledJobs = shutdownNow();
+        // Wait a while for tasks to respond to being cancelled
+        if (!awaitTermination(FORCEFUL_SHUTDOWN_TIMEOUT_SECS, SECONDS)) {
+          logger.warn("Scheduler " + this.toString() + " did not shutdown gracefully after " + gracefulShutdownTimeoutSecs
+              + " seconds. " + cancelledJobs.size() + " jobs were cancelled.");
+        } else {
+          if (!cancelledJobs.isEmpty()) {
+            logger.warn("Scheduler " + this.toString() + " terminated. " + cancelledJobs.size() + " jobs were cancelled.");
+          }
+        }
+      }
+    } catch (InterruptedException ie) {
+      // (Re-)Cancel if current thread also interrupted
+      shutdownNow();
+      // Preserve interrupt status
+      currentThread().interrupt();
+    }
+  }
+
+  @Override
+  protected <T> RunnableFuture<T> newTaskFor(Callable<T> callable) {
+    final RunnableFutureDecorator<T> decorated = new RunnableFutureDecorator<>(super.newTaskFor(callable), this);
+    currentTasks.add(decorated);
+    return decorated;
+  }
+
+  @Override
+  protected <T> RunnableFuture<T> newTaskFor(Runnable runnable, T value) {
+    final RunnableFutureDecorator<T> decorated = new RunnableFutureDecorator<>(super.newTaskFor(runnable, value), this);
+    currentTasks.add(decorated);
+    return decorated;
+  }
+
+  @Override
+  public void execute(Runnable command) {
+    checkShutdown();
+    requireNonNull(command);
+
+    executor.execute(command);
+  }
+
+  protected void checkShutdown() {
+    if (isShutdown()) {
+      throw new RejectedExecutionException(this.toString() + " already shutdown");
```

Comment:
Use a RejectedExceptionHandler, unless we are sure about hard coding AbortPolicy behaviour in our Scheduler.